File: codechef/contests/SNCKPB17/SNDISCUS.py


class Snake:

    def __init__(self, x1, y1, x2, y2):
        self.x1 = x1
        self.x2 = x2
        self.y1 = y1
        self.y2 = y2

    def horizontal(self):
        return self.y1 == self.y2

    def vertical(self):
        return self.x1 == self.x2

    def dist(self, x, y):
        if self.horizontal():
            first = abs(self.y1 - y)
            second = 0
            if min(self.x1, self.x2) > x or max(self.x1, self.x2) < x:
                second = min(abs(self.x1 - x), abs(self.x2 - x))
            return first + second
        # Now the snake has to be vertical
        first = abs(self.x1 - x)
        second = 0
        if min(self.y1, self.y2) > y or max(self.y1, self.y2) < y:
            second = min(abs(self.y1 - y), abs(self.y2 - y))
        return first + second


t = int(raw_input())
for _ in xrange(t):
    n = int(raw_input())
    snakes = []
    for i in xrange(n):
        x, y, z, w = map(int, raw_input().split())
        snakes.append(Snake(x, y, z, w))
    min_dist = 10**5
    for x in xrange(1, 51):
        for y in xrange(1, 51):
            max_dist = 0
            for snake in snakes:
                max_dist = max(max_dist, snake.dist(x, y))
            min_dist = min(min_dist, max_dist)
    print min_dist

File: codechef/contests/SNCKPB17/SNDISCUS.cc
#include <iostream>
#include <cmath>

using namespace std;

int snakes[50][4];
int n;

inline bool vertical(int i){
    return snakes[i][0] == snakes[i][2];
}

inline bool horizontal(int i){
    return snakes[i][1] == snakes[i][3];
}

inline int dist(int i, int x, int y){
    if(horizontal(i)){
        int first = abs(snakes[i][1] - y);
        int second = 0;
        if( min(snakes[i][0], snakes[i][2]) > x || 
            max(snakes[i][0], snakes[i][2]) < x){
            second = min(abs(snakes[i][0] - x), abs(snakes[i][2] - x));
        }
        return first + second;
    }
    // its gotta be vertical now
    int first = abs(snakes[i][0] - x);
    int second = 0;
    if( min(snakes[i][1], snakes[i][3]) > y ||
        max(snakes[i][1], snakes[i][3]) < y){
        second = min(abs(snakes[i][1] - y), abs(snakes[i][3] - y));
    }
    return first + second;
}

int main(void){
    ios::sync_with_stdio(false);
    int t;
    cin >> t;
    while(t--){
        cin >> n;
        for(int i = 0; i < n; i += 1){
            for(int j = 0; j < 4; j += 1){
                cin >> snakes[i][j];
            }
        }
        int min_dist = 1e5;
        for(int x = 1; x <= 50; x++){
            for(int y = 1; y <= 50; y++){
                int max_dist = 0;
                for(int i = 0; i < n; i += 1){
                    max_dist = max(max_dist, dist(i, x, y));
                }
                min_dist = min(min_dist, max_dist);
            }
        }
        cout << min_dist << endl;
    }
}